### PR TITLE
Introduce curried role binding

### PR DESCRIPTION
**Pull request by Kaiepi**
As [before](https://github.com/rakudo/rakudo/pull/4436), the following should be valid:

```raku
role Foo[Any ::T] { }
role Bar[::T] does Foo[T] { }
```

But we now also consider:

```raku
role Foo[Any ::T] { }
role Bar[::T] is Foo[T] { }
```

And:

```raku 
role OfIndex[Enumeration ::T] does T { }
```

Curries of parametric role groups already have a notion of a binding in their cached `$!candidate`. Formalizing this via new `bind` metamethods on `Metamodel::CurriedRoleHOW` and `Metamodel::ParametricRoleGroupHOW` makes it possible to fix several issues with role typechecking because if a binding is allowed to fail, it follows operations pertaining to them can follow suit.

`Metamodel::CurriedRoleHOW.bind` defers to its role's `bind` metamethod (if any, in which case it accepts the curry it's invoked from in addition to a metaobject) to produce a final parameterization:

```
bastille% raku -e 'say Blob[uint8].^bind =:= Blob.^bind: Blob[uint8]'
True
bastille% raku -e 'say Blob[uint8].^bind.HOW.^name'
Perl6::Metamodel::ParametricRoleHOW
```

There should be opportunities for optimizations depending on this here. While this passes all tests locally, this is a draft because it has changes reminiscent of those for subsets that were too breaking for v6.c. The effect of any revisions to this on performance should be dampened by the optimizations made now, but aren't made at this point to avoid making them in paranoia.

### CODE REVIEW

**Comment by Kaiepi** [comment]
The tail end of this needs more work.

File: src/Perl6/Metamodel/CurriedRoleHOW.nqp [hunk 1]
```diff
@@ -79,77 +79,94 @@ class Perl6::Metamodel::CurriedRoleHOW
...
+    # A curried role may be bound to form a role other than its origin, e.g.  a
+    # parametric role group and the candidate we're not quite ready to fetch at
+    # construction time. If we're still not ready yet whenever we need it, the
+    # binding is ourself for now. If the origin lacks a means of producing a
+    # binding, we assume it's its own by default.
+    method bind($obj) {
+        if $!binding =:= NQPMu {
+            if nqp::can($!curried_role.HOW, 'bind') {
+                my $binding := $!curried_role.HOW.bind($!curried_role, $obj);
+                $!binding := $binding unless nqp::decont($binding) =:= $obj;
```
Comment (vrurg) [inline src/Perl6/Metamodel/CurriedRoleHOW.nqp:91]: If you ever expect `bind` to return a container (which is, perhaps, expected for Raku-land implementations of the method), then deconting must take place the line above or you end up with a container in `$!binding`.

File: src/Perl6/Metamodel/CurriedRoleHOW.nqp [hunk 2]
```diff
@@ -79,77 +79,94 @@ class Perl6::Metamodel::CurriedRoleHOW
...
+    }
 
-            self.set_language_revision($obj, $candidate-how.language-revision($!candidate));
+    method parameterize_roles($obj) {
+        if self.archetypes.generic {
+            $!is_complete := 1;
+        }
+        else {
+            my $binding := self.bind($obj);
+            unless nqp::decont($binding) =:= $obj {
```
Comment (vrurg) [inline src/Perl6/Metamodel/CurriedRoleHOW.nqp:109]: Same as above, better assume that deconting is not needed.

File: src/Perl6/Metamodel/CurriedRoleHOW.nqp [hunk 3]
```diff
@@ -79,77 +79,94 @@ class Perl6::Metamodel::CurriedRoleHOW
...
-        for @!pos_args {
-            nqp::push(@pos_args, $_);
+    # A curried role may be bound to form a role other than its origin, e.g.  a
+    # parametric role group and the candidate we're not quite ready to fetch at
+    # construction time. If we're still not ready yet whenever we need it, the
+    # binding is ourself for now. If the origin lacks a means of producing a
+    # binding, we assume it's its own by default.
+    method bind($obj) {
+        if $!binding =:= NQPMu {
+            if nqp::can($!curried_role.HOW, 'bind') {
```
Comment (vrurg) [inline src/Perl6/Metamodel/CurriedRoleHOW.nqp:89]: I'd rather make method `bind` mandatory on all role HOWs, except for `ConcreteRoleHOW` which is impossible candidate for currying. Then this condition and it's `else` block can be removed. `bind` on `ParametricRoleHOW` would then simply return itself.

File: src/Perl6/Metamodel/CurriedRoleHOW.nqp [hunk 4]
```diff
@@ -79,77 +79,94 @@ class Perl6::Metamodel::CurriedRoleHOW
...
-            my $candidate-how := $!candidate.HOW;
+        else {
+            $!binding
+        }
+    }
 
-            self.set_language_revision($obj, $candidate-how.language-revision($!candidate));
+    method parameterize_roles($obj) {
+        if self.archetypes.generic {
+            $!is_complete := 1;
```
Comment (vrurg) [inline src/Perl6/Metamodel/CurriedRoleHOW.nqp:105]: It is kind of a good idea for optimization, but it breaks typechecking of partially parameterized roles:

```raku
role R1[::T] { }
role R2[::T1, ::T2] does R1[::T1] { } 
role R3[::T] does R2[Int, T] { }
say R3.^candidates.head ~~ R1[Int];
```

This is _True_ on the current master but _False_ on your branch. Perhaps this makes sense when all arguments are generic, but what if there are too many of them?

File: src/Perl6/Metamodel/CurriedRoleHOW.nqp [hunk 5]
```diff
@@ -79,77 +79,94 @@ class Perl6::Metamodel::CurriedRoleHOW
...
-                    my $target-name := $obj.HOW.name($obj);
-                    my $role-name := $role.HOW.name($role);
-                    Perl6::Metamodel::Configuration.throw_or_die(
-                        'X::Composition::NotComposable',
-                        $role-name ~ " is not composable, so " ~ $target-name ~ " cannot compose it",
-                        :$target-name,
-                        composer => $role,
-                    )
+                if $error {
+                    # XXX: Could use a wrapper exception.
```
Comment (vrurg) [inline src/Perl6/Metamodel/CurriedRoleHOW.nqp:120]: _Must_ use. We must depart from using anonymous adhocs.

File: src/Perl6/Metamodel/CurriedRoleHOW.nqp [hunk 6]
```diff
@@ -205,43 +226,43 @@ class Perl6::Metamodel::CurriedRoleHOW
...
     }
 
     method accepts_type($obj, $checkee) {
-        # First, we locate candidate curryings to check against. If
-        # the checkee is itself a curried role, it also goes in. Note
-        # that we only want those that have the same parametric role
-        # as us.
+        # Generics' incapacity for typechecking propagates.
+        if self.archetypes.generic {
+            return 0;
```
Comment (vrurg) [inline src/Perl6/Metamodel/CurriedRoleHOW.nqp:246]: Same note about partially parameterized cases.

File: src/Perl6/Metamodel/ParametricRoleGroupHOW.nqp [hunk 7]
```diff
@@ -90,6 +90,20 @@ class Perl6::Metamodel::ParametricRoleGroupHOW
         $curried
     }
 
+    method bind($obj, $curried) {
+        my $candidate;
+        my %named_args := $curried.HOW.role_named_arguments($curried);
+        my @pos_args := [$obj];
+        nqp::splice(@pos_args, $curried.HOW.role_arguments($curried), +@pos_args, 0);
```
Comment (vrurg) [inline src/Perl6/Metamodel/ParametricRoleGroupHOW.nqp:97]: BTW, why do you use `nqp::splice` instead of `nqp::push`?

Comment (Kaiepi) [reply src/Perl6/Metamodel/ParametricRoleGroupHOW.nqp:97]: Splicing at the end of an array like I do here seems to run faster than pushing from a `for`. Elsewhere, there's usually a need to map over the arguments, in which case I'd expect the loop to be more efficient instead.

File: src/Perl6/Metamodel/CurriedRoleHOW.nqp (hunk 3 above)
Comment (Kaiepi) [reply src/Perl6/Metamodel/CurriedRoleHOW.nqp:89]: I thought the overhead of the identity binding for `Metamodel::ParametricRoleHOW` would be greater than the `nqp::can`s, but that doesn't seem to be the case.

File: src/Perl6/Metamodel/ParametricRoleGroupHOW.nqp (hunk 7 above)
Comment (vrurg) [reply src/Perl6/Metamodel/ParametricRoleGroupHOW.nqp:97]: Makes sense. I forgot that `push` is for single items only.

**Comment by Kaiepi** [comment]
Bar the ad-hoc error, the issues brought up so far should be addressed. The strategy to optimizing role typechecking is now to eliminate the recursion from the doee typechecks. If that's not reasonable to do, some of the performance gains might be salvageable with fast/slow paths for doee typechecking.

The following should also be taken into consideration now, but aren't at this point:

```raku
enum Index <A B C>;

class Foo does Index { }
Foo ~~ Index;
```

```raku
class Foo does Buf[uint8] { }
Foo ~~ Buf;
Foo ~~ Buf[uint8];

class Bar does Buf { }
Bar ~~ Buf;
Bar ~~ Buf[uint8];
```

These should share a common theme in their solution.

File: src/Perl6/Metamodel/CurriedRoleHOW.nqp [hunk 8]
```diff
@@ -79,28 +79,36 @@ class Perl6::Metamodel::CurriedRoleHOW
...
-        nqp::push(@pos_args, $obj);
-        for @!pos_args {
-            nqp::push(@pos_args, $_);
+    # A curried role may be bound to form a role other than its origin, e.g.  a
+    # parametric role group and the candidate we're not quite ready to fetch at
+    # construction time. If we're still not ready yet whenever we need it, the
+    # binding is ourself for now.
+    method bind($obj) {
+        if $!binding =:= NQPMu {
+            my $binding := nqp::decont($!curried_role.HOW.bind($!curried_role, $obj));
```
Comment (vrurg) [inline src/Perl6/Metamodel/CurriedRoleHOW.nqp:88]: I see two inconsistencies here. 

First, you still use deconting with `bind`, but it's the only place where you do so. I think you can safely throw it away.

The other one worries me more. Your `bind` methods has different signatures. It means that the code invoking the method must know exactly what kind of role it currently works with. Would it always be the case? I wouldn't rely on this. So, the second parameter must be introduced but made optional where it's not used.

**Comment by Kaiepi** [comment]
Closing in favour of https://github.com/rakudo/rakudo/pull/4478.

**Pull request closed without merging**